ntoskrnl/ke/i386/mp.c

```diff
@@ -0,0 +1,140 @@
+    do
+    {
+        ProcessorCount++;
+
+        /* Attempt to allocate memory for its new setup */
+        SIZE_T APInfoSize = sizeof(struct APInfo) + 2;
+        PVOID PAPInfo = MmAllocateContiguousMemory(APInfoSize, HighestPhysicalAddress);
+
+        if (!PAPInfo)
+        {
+            ASSERT("KeStartAllProcessors: Memory Allocation has failed");
+        }
+
+        RtlCopyMemory((PVOID)((ULONG_PTR)PAPInfo + sizeof(KPCR) + sizeof(KTSS)), (PVOID)GdtDesc.Base, GdtDesc.Limit + 1);
+        RtlCopyMemory((PVOID)((ULONG_PTR)PAPInfo + sizeof(KPCR) + sizeof(KTSS) + GdtDesc.Limit + 1), (PVOID)IdtDesc.Base, IdtDesc.Limit + 1);
+
+        KPROCESSOR_STATE ProcessorState;
+        
+        ProcessorState.SpecialRegisters.Gdtr.Limit = GdtDesc.Limit;
+        ProcessorState.SpecialRegisters.Gdtr.Base = (ULONG)((ULONG_PTR)PAPInfo + sizeof(KPCR) + sizeof(KTSS));
+        ProcessorState.SpecialRegisters.Idtr.Limit = IdtDesc.Limit;
+        ProcessorState.SpecialRegisters.Idtr.Base = (ULONG)((ULONG_PTR)PAPInfo + sizeof(KPCR) + sizeof(KTSS) + GdtDesc.Limit + 1);
+
+        /* Prep Cr Regsters */
+        ProcessorState.SpecialRegisters.Cr0 = __readcr0();
+        ProcessorState.SpecialRegisters.Cr2 = __readcr2();
+        ProcessorState.SpecialRegisters.Cr3 = __readcr3();
+        ProcessorState.SpecialRegisters.Cr4 = __readcr4();
+
+          /* Prepare Segment Registers */
+        ProcessorState.ContextFrame.SegCs = KGDT_R0_CODE;
+        ProcessorState.ContextFrame.SegSs = KGDT_R0_DATA;
+        ProcessorState.ContextFrame.SegDs = KGDT_R0_DATA;
+        ProcessorState.ContextFrame.SegEs = KGDT_R0_DATA;
+
+        /* Clear GS */
+        ProcessorState.ContextFrame.SegGs = 0;
+        /* Set FS to PCR */
+        ProcessorState.ContextFrame.SegFs = KGDT_R0_PCR;
+
+        /* Load TSR */
+        ProcessorState.SpecialRegisters.Tr = KGDT_TSS;
+
+        ProcessorState.ContextFrame.Eip = (ULONG_PTR)KiSystemStartup;
+
+        /* Prep GDT and IDT based on BSPs */
+        PKGDTENTRY pGdt = (PKGDTENTRY)((ULONG_PTR)PAPInfo + sizeof(KPCR) + sizeof(KTSS));
+        PKIDTENTRY pIdt = (PKIDTENTRY)((ULONG_PTR)PAPInfo + sizeof(KPCR) + sizeof(KTSS) + GdtDesc.Limit + 1);
+
+        /* Allocate PCR - TSS - InitalThread */
+        PKPCR pKPcr = (PKPCR)PAPInfo;
+        PKTSS pKTss = (PKTSS)((ULONG_PTR)PAPInfo + sizeof(KPCR));
+        PKTHREAD InitialThread = (PKTHREAD)((ULONG_PTR) PAPInfo + sizeof(KPCR) + sizeof(KTSS));
+
+        /* Prep some stacks for the AP */
+        PVOID KernelStack;
+        PVOID DPCStack;
+
+        KernelStack = MmCreateKernelStack(FALSE, 0);
+        if (!KernelStack)
+        {
+            ASSERT("KeStartAllProcessors: MmCreateKernelStack has failed for an AP");
+        }
+
+        DPCStack = MmCreateKernelStack(FALSE, 0);
+        if (!DPCStack)
+        {
+            MmDeleteKernelStack(KernelStack, FALSE);
+        }
+
+        /* Initalize a new PCR for the specific AP */
+        KiInitializePcr(ProcessorCount,
```

Comment:
Why does this need to be initialized? What's wrong with doing that in KiSystemStartup, like done for the BSP?

Reply:
https://github.com/reactos/reactos/blob/master/ntoskrnl/ke/i386/kiinit.c#L755-L756

Reply:
Windows does it like that too.

Reply:
Seems a bit strange, but ok.